server/src/http/feature_refresher.rs

```diff
@@ -0,0 +1,467 @@
+                                        debug!("No update needed. Will update last check time");
+                                        self.update_last_check(&refresh.token.clone());
+                                    }
+                                    ClientFeaturesResponse::Updated(features, etag) => {
+                                        debug!("Got updated client features. Updating features");
+                                        let key = cache_key(refresh.token.clone());
+                                        self.update_last_refresh(&refresh.token, etag);
+                                        self.features_cache.entry(key.clone()).and_modify(|existing_data| {
+                                            *existing_data = existing_data.clone().upsert(features.clone());
+                                        }).or_insert(features.clone());
+                                        if self.engine_cache.contains_key(&key) {
+                                            self.engine_cache.entry(key.clone()).and_modify(|engine| {
+                                                if let Some(f) = self.features_cache.get(&key) {
+                                                        let mut new_state = EngineState::default();
+                                                        new_state.take_state(f.clone());
+                                                        *engine = new_state;
+                                                }
+                                            });
+                                        } else {
+                                            let mut new_state = EngineState::default();
+                                            new_state.take_state(features.clone());
+                                            self.engine_cache.insert(key, new_state);
+                                        }
+                                        if let Some(persist) = self.persistence.clone() {
+                                            let feature_save = persist.save_features(
+                                                self.features_cache
+                                                    .iter()
+                                                    .map(|e| (e.key().clone(), e.value().clone()))
+                                                    .collect());
+
+                                            let refresh_targets_save = persist.save_refresh_targets(self.tokens_to_refresh.iter().map(|e| e.value().clone()).collect());
+                                            let _ = futures::future::join_all(vec![feature_save, refresh_targets_save]).await;
+                                        }
+                                    }
+                                },
+                                Err(e) => {
+                                    warn!("Couldn't refresh features: {e:?}");
+                                }
```

Comment:
Not quite sure what you mean here?

This is the logic for keeping our features up to date from Unleash, that call to Unleash yields a list of features. If Unleash doesn't return a useful response there isn't much we can do outside of waiting for a bit and trying again.

Or are you saying in case of bad auth, we should invalidate the token we're using to refresh?

Reply:
Let me place a scenario:
1. Unleash has toggles: `a, b, c`
2. Edge replicates toggles: `a, b, c`
3. Unleash removes toggle `a`
4. Client requests toggle `a` to Edge

The expected result should be as if the toggle `a` does not exist. Eventual consistency is fine. If we swap data cached with fresh data, this will work, but if we iterate and update `a` might remain in edge forever. I'm just trying to confirm we're not in the last situation